crates/bevy_render/src/renderer/mod.rs

```diff
@@ -345,25 +360,87 @@ impl RenderContext {
         TrackedRenderPass::new(&self.render_device, render_pass)
     }
 
-    /// Append a [`CommandBuffer`] to the queue.
+    /// Append a [`CommandBuffer`] to the command buffer queue.
     ///
     /// If present, this will flush the currently unflushed [`CommandEncoder`]
-    /// into a [`CommandBuffer`] into the queue before append the provided
+    /// into a [`CommandBuffer`] into the queue before appending the provided
     /// buffer.
     pub fn add_command_buffer(&mut self, command_buffer: CommandBuffer) {
         self.flush_encoder();
-        self.command_buffers.push(command_buffer);
+
+        self.command_buffer_queue
+            .push(QueuedCommandBuffer::Ready(command_buffer));
     }
 
-    /// Finalizes the queue and returns the queue of [`CommandBuffer`]s.
+    /// Append a function that will generate a [`CommandBuffer`] to the
+    /// command buffer queue, to be ran later.
+    ///
+    /// If present, this will flush the currently unflushed [`CommandEncoder`]
+    /// into a [`CommandBuffer`] into the queue before appending the provided
+    /// buffer.
+    pub fn add_command_buffer_generation_task(
+        &mut self,
+        task: impl FnOnce(RenderDevice) -> CommandBuffer + 'w + Send,
+    ) {
+        self.flush_encoder();
+
+        self.command_buffer_queue
+            .push(QueuedCommandBuffer::Task(Box::new(task)));
+    }
+
+    /// Finalizes and returns the queue of [`CommandBuffer`]s.
+    ///
+    /// This function will wait until all command buffer generation tasks are complete
+    /// by running them in parallel (where supported).
     pub fn finish(mut self) -> Vec<CommandBuffer> {
         self.flush_encoder();
-        self.command_buffers
+
+        if self.command_buffer_queue.is_empty() {
+            self.command_buffer_queue
```

Comment:
This is being done, even though the queue is empty? Shouldn't we just return `Vec::new()`?

Reply:
I don't know why the code was like that. I removed all the redundant stuff.

Reply:
I actually looked at the generated assembly of this, and for whatever reason the original compiled into fewer operations.